urlcdab7d8364_jrwalsh_CycTools_tgz-pJ8-edu_iastate_cyctools_MainJ8/src/edu/iastate/cyctools/CycDataBaseAccess.java

```diff
@@ -0,0 +1,687 @@
+			conn.revertKB();
+		} catch (PtoolsErrorException e) {
+			e.printStackTrace();
+		}
+	}
+	
+	public void commitFrameUpdates(ArrayList<AbstractFrameEdit> frameUpdates) {
+		for (AbstractFrameEdit frameUpdate : frameUpdates) {
+			try {
+				boolean result = frameUpdate.commit(conn);
+				if (!result) System.err.println("Commit failed!");
+			} catch (PtoolsErrorException e) {
+				e.printStackTrace();
+			}
+		}
+	}
+	
+	public Frame updateLocalFrame(String frameID, ArrayList<AbstractFrameEdit> frameUpdates) {
+		Frame frame = null;
+		try {
+			frame = Frame.load(conn, frameID);
+			frame.update();
+			for (AbstractFrameEdit frameUpdate : frameUpdates) {
+				frameUpdate.commitLocal(frame);
+			}
+		} catch (PtoolsErrorException e) {
+			e.printStackTrace();
+		}
+		return frame;
+	}
+	
+	public DefaultMutableTreeNode frameToTree(String frameID) {
+ 		Frame frame = null;
+ 		try {
+			frame = Frame.load(conn, frameID);
+		} catch (PtoolsErrorException e) {
+			e.printStackTrace();
+			return null;
+		}
+ 		
+ 		DefaultMutableTreeNode root = null;
+ 		try {
+ 			root = createNodesFromFrame(new DefaultMutableTreeNode(frame.getLocalID()), frame);
+		} catch (PtoolsErrorException e) {
+			e.printStackTrace();
+		}
+ 		
+ 		return root;
+ 	}
+ 	
+	private DefaultMutableTreeNode createNodesFromFrame(DefaultMutableTreeNode top, Frame frame) throws PtoolsErrorException {
+ 		// Get slot names
+		HashMap<String, ArrayList> slots = frame.getSlots();
+ 		for (String slotLabel : slots.keySet()) {
+			DefaultMutableTreeNode slotNode = new DefaultMutableTreeNode(slotLabel);
+ 			// Get slot values
+			ArrayList<String> slotValues = new ArrayList<String>();
+ 			for (Object slotValue : conn.getSlotValues(frame.getLocalID(), slotLabel)) {
+ 				if (slotValue instanceof String) slotValues.add((String) slotValue);
+ 				else slotValues.add(JavacycConnection.ArrayList2LispList((ArrayList) slotValue));
+ 			}
+ 			for (String slotValue : slotValues) {
+ 					DefaultMutableTreeNode slotValueNode = new DefaultMutableTreeNode(slotValue);
+ 	 				// Get annotation names
+ 	 				ArrayList<String> annotLabels = frame.getAllAnnotLabels(slotLabel, slotValue);
+ 	 				for (String annotLabel : annotLabels) {
+ 	 					DefaultMutableTreeNode annotLabelNode = new DefaultMutableTreeNode(annotLabel);
+ 	 					// Get annotation values
+ 	 					ArrayList<String> annotValues = (ArrayList<String>) frame.getAnnotations(slotLabel, slotValue, annotLabel);
+ 	 					for (String annotValue: annotValues) {
+ 	 						DefaultMutableTreeNode annotValueNode = new DefaultMutableTreeNode(annotValue);
+ 	 						annotLabelNode.add(annotValueNode);
+ 	 					}
+ 	 					slotValueNode.add(annotLabelNode);
+ 	 				}
+ 	 				slotNode.add(slotValueNode);
+ 			}
+ 			top.add(slotNode);
+ 		}
+ 		return top;
+ 	}
+	
+	public DefaultTableModel getSearchResultsTable(String text, String type) {
+		progressMonitor = new ProgressMonitor(DefaultController.mainJFrame, "Running a Long Task", "", 0, 100);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code assigns a resource to a class field and there is no method that closes or returns this resource. This can result in a resource leak that slows down or crashes your system.

**Fix**
Create a method in this class that closes the following resource: *progressMonitor*. After you use this resource, use this method to prevent a resource leak.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).